### PR TITLE
dialects: (builtin) constrain DensIntOrFPElementsattr to the correct nb of elements

### PR DESCRIPTION
**Pull request by jorendumoulin**

### CODE REVIEW

File: xdsl/dialects/builtin.py [hunk 1]
```diff
@@ -2040,25 +2043,38 @@ def from_list(
...
             new_type = cast(RankedStructure[IndexType], type)
-            new_data = cast(Sequence[int] | Sequence[IntegerAttr[IndexType]], data)
+            new_data = cast(Sequence[int] | Sequence[IntegerAttr[IndexType]], new_data)
             return DenseIntOrFPElementsAttr.create_dense_index(new_type, new_data)
 
     @staticmethod
     def vector_from_list(
         data: Sequence[int] | Sequence[float],
         data_type: IntegerType | IndexType | AnyFloat,
+        shape: Sequence[int],
```
Comment (superlopuh) [inline xdsl/dialects/builtin.py:2079]: I still think that we should not make a breaking change here, instead allowing to pass a custom shape if wanted

File: xdsl/dialects/builtin.py [hunk 2]
```diff
@@ -1944,6 +1944,9 @@ def get_shape(self) -> tuple[int, ...]:
     def get_element_type(self) -> IntegerType | IndexType | AnyFloat:
         return self.type.get_element_type()
 
+    def __len__(self) -> int:
+        return prod(self.get_shape())
```
[the remark this answers is not in the document]
Comment (superlopuh) [reply xdsl/dialects/builtin.py:1948]: Why not this?

File: xdsl/dialects/builtin.py (hunk 1 above)
Comment (jorendumoulin) [reply xdsl/dialects/builtin.py]: I removed it for now

Comment (superlopuh) [reply xdsl/dialects/builtin.py]: I feel like there's been a miscommunication, this is what I was recommending:
```
        shape: Sequence[int] | None = None,
```

Comment (superlopuh) [reply xdsl/dialects/builtin.py]: That way you can provide the shape, or it can be inferred from the length of the data

**Comment by superlopuh** [comment]
Could you please assign yourself and add the appropriate label?

File: xdsl/dialects/builtin.py [hunk 3]
```diff
@@ -2040,17 +2043,33 @@ def from_list(
         ),
         data: Sequence[int | float] | Sequence[AnyIntegerAttr] | Sequence[AnyFloatAttr],
     ) -> DenseIntOrFPElementsAttr:
+        # zero rank type should only hold 1 value
+        if not type.get_shape() and len(data) != 1:
+            raise ValueError(
+                f"A zero-rank {type.name} can only hold 1 value but {len(data)} were given."
+            )
```
Comment (superlopuh) [inline xdsl/dialects/builtin.py:2050]: Could you please add a test for this?

File: xdsl/dialects/builtin.py [hunk 4]
```diff
@@ -1944,6 +1944,9 @@ def get_shape(self) -> tuple[int, ...]:
     def get_element_type(self) -> IntegerType | IndexType | AnyFloat:
         return self.type.get_element_type()
 
+    def __len__(self) -> int:
+        return len(self.data)
+
```
Comment (superlopuh) [inline xdsl/dialects/builtin.py:1949]: could you please add a tiny test for this?

File: xdsl/dialects/builtin.py (hunk 3 above)
Comment (superlopuh) [reply xdsl/dialects/builtin.py:2050]: My b

File: tests/dialects/test_builtin.py [hunk 5]
```diff
@@ -264,10 +264,29 @@ def test_DenseIntOrFPElementsAttr_fp_type_conversion():
...
+    assert attr.type == AnyTensorType(f32, [1])
+
+    # legal normal tensor
+    attr = DenseIntOrFPElementsAttr.tensor_from_list([5.5, 5.6], f32, [2])
+    assert attr.type == AnyTensorType(f32, [2])
+
+    # splat initialization
+    attr = DenseIntOrFPElementsAttr.tensor_from_list([4], f32, [4])
+    assert attr.type == AnyTensorType(f32, [4])
+    assert tuple(attr.get_values()) == (4, 4, 4, 4)
```
Comment (superlopuh) [inline tests/dialects/test_builtin.py:288]:
```suggestion
    # legal 1 element tensor
    attr = DenseIntOrFPElementsAttr.tensor_from_list([5.5], f32, [1])
    assert attr.type == AnyTensorType(f32, [1])
    assert len(attr) == 1

    # legal normal tensor
    attr = DenseIntOrFPElementsAttr.tensor_from_list([5.5, 5.6], f32, [2])
    assert attr.type == AnyTensorType(f32, [2])
    assert len(attr) == 1

    # splat initialization
    attr = DenseIntOrFPElementsAttr.tensor_from_list([4], f32, [4])
    assert attr.type == AnyTensorType(f32, [4])
    assert tuple(attr.get_values()) == (4, 4, 4, 4)
    assert len(attr) == 4
```

**Review by superlopuh — APPROVED** [review]

File: xdsl/dialects/builtin.py (hunk 1 above)
Comment (jorendumoulin) [reply xdsl/dialects/builtin.py]: oh ok yes that makes sense

**Pull request merged**